server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2690,17 +2689,15 @@ public void maybeSyncGlobalCheckpoint(final String reason) {
         final SeqNoStats stats = getEngine().getSeqNoStats(replicationTracker.getGlobalCheckpoint());
         final boolean asyncDurability = indexSettings().getTranslogDurability() == Translog.Durability.ASYNC;
         if (stats.getMaxSeqNo() == stats.getGlobalCheckpoint() || asyncDurability) {
-            final ObjectLongMap<String> globalCheckpoints = getInSyncGlobalCheckpoints();
-            final long globalCheckpoint = replicationTracker.getGlobalCheckpoint();
+            final var trackedGlobalCheckpointsNeedSync = replicationTracker.trackedGlobalCheckpointsNeedSync();
```

Comment:
Wouldn't it make sense to keep this in line in the conditional since it's an `OR` so maybe we don't need to evaluate it all the time?

Reply:
Ehh maybe, I didn't want to have to reason about whether the ordering between this and checking `replication tracker.pendingInSync()` was important. Still the expensive bit is `getInSyncGlobalCheckpoints()` which we did every time anyway, and `asyncDurability` is almost always false so I doubt it makes much difference.